pkg/verify/verify.go

```diff
@@ -0,0 +1,203 @@
+// verify the root hash via VerifyRootHash.
+//nolint
+func VerifyInclusion(ctx context.Context, e *models.LogEntryAnon) error {
+	if e.Verification == nil || e.Verification.InclusionProof == nil {
+		return errors.New("inclusion proof not provided")
+	}
+
+	hashes := [][]byte{}
+	for _, h := range e.Verification.InclusionProof.Hashes {
+		hb, _ := hex.DecodeString(h)
+		hashes = append(hashes, hb)
+	}
+
+	rootHash, err := hex.DecodeString(*e.Verification.InclusionProof.RootHash)
+	if err != nil {
+		return err
+	}
+
+	// Verify the inclusion proof.
+	entryBytes, err := base64.StdEncoding.DecodeString(e.Body.(string))
+	if err != nil {
+		return err
+	}
+	leafHash := rfc6962.DefaultHasher.HashLeaf(entryBytes)
+
+	if err := proof.VerifyInclusion(rfc6962.DefaultHasher, uint64(*e.Verification.InclusionProof.LogIndex),
+		uint64(*e.Verification.InclusionProof.TreeSize), leafHash, hashes, rootHash); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// VerifySignedEntryTimestamp verifies the entry's SET against the provided
+// public key.
+//nolint
+func VerifySignedEntryTimestamp(ctx context.Context, e *models.LogEntryAnon, verifier signature.Verifier) error {
+	if e.Verification == nil {
+		return fmt.Errorf("missing verification")
+	}
+	if e.Verification.SignedEntryTimestamp == nil {
+		return fmt.Errorf("signature missing")
+	}
+
+	type bundle struct {
+		Body           interface{} `json:"body"`
+		IntegratedTime int64       `json:"integratedTime"`
+		// Note that this is the virtual index.
+		LogIndex int64  `json:"logIndex"`
+		LogID    string `json:"logID"`
+	}
+	bundlePayload := bundle{
+		Body:           e.Body,
+		IntegratedTime: *e.IntegratedTime,
+		LogIndex:       *e.LogIndex,
+		LogID:          *e.LogID,
+	}
+	contents, err := json.Marshal(bundlePayload)
+	if err != nil {
+		return fmt.Errorf("marshaling bundle: %w", err)
+	}
+	canonicalized, err := jsoncanonicalizer.Transform(contents)
+	if err != nil {
+		return fmt.Errorf("canonicalizing bundle: %w", err)
+	}
+
+	// verify the SET against the public key
+	if err := verifier.VerifySignature(bytes.NewReader(e.Verification.SignedEntryTimestamp),
+		bytes.NewReader(canonicalized), options.WithContext(ctx)); err != nil {
+		return fmt.Errorf("unable to verify bundle: %w", err)
+	}
+	return nil
+}
+
+// VerifyLogEntry performs verification of a LogEntry given a Rekor verifier.
+// Performs inclusion proof verification up to a verified root hash and
+// SignedEntryTimestamp verification.
+//nolint
+func VerifyLogEntry(ctx context.Context, rClient *client.Rekor, e *models.LogEntryAnon, verifier signature.Verifier) error {
+	// Verify the inclusion proof.
+	if err := VerifyInclusion(ctx, e); err != nil {
+		return err
+	}
+
+	// Verify the root hash.
+	rootHash, err := hex.DecodeString(*e.Verification.InclusionProof.RootHash)
+	if err != nil {
+		return err
+	}
+	tid := fmt.Sprintf("%x", *e.LogID)
+
+	if err := VerifyRootHash(ctx, rClient, verifier,
```

Comment:
Messaged offline about this, but I don't think this is necessary. Consistency should be verified from a known trusted state I believe. Otherwise, you're effectively asking the log for both ends of the consistency proof, so it could forge it.

I think this function should optionally take in a signed tree head to verify consistency from it to the root hash returned by the inclusion proof.

Reply:
I think this is why we never added it in cosign. Adding this check is essentially only to prevent split tree attacks, and mitigation for that requires storing client state in between invocations.

I'm +1 on doing this (and adding the library function is fine), but the change in cosign will be a bit more invasive if we want to make it useful.

Reply:
Yea, consistency checks are for enforcing the append-only property of a log, and would require client state. I think the function should support such a verification, and it could be added to cosign as an optional online lookup.

Split view attacks would require a gossip protocol which is out of scope for this.

Reply:
Yeah, I made all the helper methods here public so people could pick and choose.

However, note there's not point in verifying an inclusion proof without some STH verification. So TBH, cosign should only verify the SET

Reply:
I think that's correct - we could leave verification of the inclusion proof to the signer.

Reply:
> Split view attacks would require a gossip protocol which is out of scope for this.

I think if clients store state you still get protection, it's just that the gossip protocol is manual and a Twitter it a Github issue.

Reply:
OK to resolve this: I'll keep `VerifyInclusion` and `VerifySignedEntryTimestamp` in `VerifyLogEntry`. 
Once https://github.com/sigstore/rekor/issues/988 is resolved, then we can add verification of the STH given in the return response against an optional old state in there too. TODO added.

I'll re-name `VerifyRootHash` to `VerifySTH`: that verifies a given STH against a newly fetched STH (by a consistency proof)